### DISCUSSION TITLE
Velero with Longhorn CSI snapshot support

### DISCUSSION THREAD

**Discussion by ehemmerlin** [longhorn/longhorn#8801] [Q&A] [answered]
We are currently using Velero's Container Storage Interface (CSI) snapshot support and it's CSI volume data movement feature. connected to Longhorn storage which supports CSI snapshots.

The issue we face is that the backups use a lot of disk space, as Longhorn replicates the data of each snapshot the same way it does for all of our Kubernetes volumes. Moreover it retains these volumes, as it's the default behaviour of our storage class, so volumes created during backups are never deleted even if the backup has expired and no longer exists.

We are looking for a way to remove the volumes created by the snapshots after the backup is deleted: is there a way to specify a different storage class for Velero's backup in Longhorn (we could create one with a Delete reclaim policy instead of a Retain one) or another way to solve this issue and delete the volumes from expired backups?

**Comment by innobead** [comment, marked as answer]
> We are looking for a way to remove the volumes created by the snapshots after the backup is deleted: is there a way to specify a different storage class for Velero's backup in Longhorn (we could create one with a Delete reclaim policy instead of a Retain one) or another way to solve this issue and delete the volumes from expired backups?

Did you mean the flow like below? Or, please help clarify it. 

- volume -> create a VolumeSnapshot -> data movement by restoring a temporary volume from the VolumeSnapshot, then move the data to a backup storage-> delete the temporary volume and the volume snapshot

When doing the VolumeSnapshot, what type of snapshot did you use? [`external backup`](https://longhorn.io/docs/1.6.2/snapshots-and-backups/csi-snapshot-support/csi-volume-snapshot-associated-with-longhorn-backup/) or [`in-cluster snapshot`](https://longhorn.io/docs/1.6.2/snapshots-and-backups/csi-snapshot-support/csi-volume-snapshot-associated-with-longhorn-snapshot/)? Longhorn supports VolumeSnapshot to create a local snapshot or remote backup. 

If it's a remote backup, in the current versions, when a backup is deleted, it will be removed from the backup storage, but the corresponding in-cluster snapshot for the volume will not be. We have a ticket for this improvement at https://github.com/longhorn/longhorn/issues/8365.

cc @c3y1huang @FrankYang0529 @derekbit

**Reply by kamalakar22** [reply]
I'm working on setting up Velero for backup and restore in a Kubernetes environment where I'm using Longhorn as the default storage class. The goal is to back up PVC data using CSI snapshots with Velero and then restore that data in a different Kubernetes cluster. Additionally, I'm using MinIO as my S3-compatible storage for backups.

Here's the configuration I'm aiming to set up:

Kubernetes Version: v1.28.6+rke2r1 Velero Version: v1.14.0 Storage Class: longhorn driver.longhorn.io Backup Storage: MinIO

I have tried the following steps but have encountered some challenges:

Installed Velero using MinIO as the backup storage: velero install
--provider aws
--bucket
--plugins velero/velero-plugin-for-aws:v1.2.0,velero/velero-plugin-for-csi:v0.2.0
--backup-location-config region=minio,s3ForcePathStyle=true,s3Url=http://:9000
--secret-file ./minio-credentials
--snapshot-location-config region=minio Added the CSI snapshot support:

Enabled --use-volume-snapshots=true when setting up Velero. Configured velero-plugin-for-csi. Backed up PVCs using: velero backup create --include-resources pvc,pv

However, I'm having trouble configuring Longhorn's CSI snapshot support correctly to ensure the PVC data is backed up properly and can be restored in another cluster. I also want to make sure the snapshot can be used in the destination cluster for PVC restoration.

Questions:

How can I configure Velero to work with Longhorn using the CSI snapshot for PVCs and MinIO as the backup storage? What are the best practices for restoring backups in another cluster with Longhorn StorageClass and MinIO storage? Any additional settings or plugins required to ensure smooth migration across clusters?

Additional Information: Both clusters have Longhorn installed and configured similarly. I am using MinIO as my S3-compatible bucket storage, and it is working well with Velero, but the issue seems to arise when restoring PVC data in a different cluster. Any help or insights would be appreciated. Thanks!

**Reply by ehemmerlin** [reply]
Hello @kamalakar22, FYI by following [this post](https://picluster.ricsanfre.com/docs/backup/#enable-csi-snapshots-support-in-k3s) we were able to configure Velero to work with Longhorn using the CSI snapshot for PVCs and MinIO as the backup storage. Nevertheless I encourage you to upgrade Velero to version 1.14.1 as it fixed some issues we had using CSI snapshots  and Data Mover with Longhorn and Minio.

**Comment by ehemmerlin** [comment]
Thanks for your answer @innobead and congratulations for the good work Longhorn is doing for us. Yes, your workflow seems to be what we are doing. The closest post describing our Velero/Longhorn/Minio architecture is this one (indeed without Raspberry PI): https://picluster.ricsanfre.com/docs/backup/#enable-csi-snapshots-support-in-k3s
![image](https://github.com/user-attachments/assets/5a3f767d-a13b-4bad-96b4-1c2f1c5e3329)

When doing the VolumeSnapshot, we use a snapshot of type [external backup](https://longhorn.io/docs/1.6.2/snapshots-and-backups/csi-snapshot-support/csi-volume-snapshot-associated-with-longhorn-backup/)

**Comment by ehemmerlin** [comment]
Longhorn and Velero team did a great job on this, that fixed this issue. As VolumeSnapshots created by the Velero CSI plugins are retained only for the lifetime of the backup even if the DeletionPolicy on the VolumeSnapshotClass is set to Retain, we don't need anymore to set a different backup storage class for data movement backups. This issue can be closed.

> VolumeSnapshots created by the Velero CSI plugins are retained only for the lifetime of the backup even if the DeletionPolicy on the VolumeSnapshotClass is set to Retain. To accomplish this, during deletion of the backup the prior to deleting the VolumeSnapshot, VolumeSnapshotContent object is patched to set its DeletionPolicy to Delete. Deleting the VolumeSnapshot object will result in cascade delete of the VolumeSnapshotContent and the snapshot in the storage provider.

See more here: https://velero.io/docs/main/csi/#implementation-choices
